src/Interpreters/Squashing.cpp

```diff
@@ -0,0 +1,165 @@
+#include <vector>
+#include <Interpreters/Squashing.h>
+#include <Common/CurrentThread.h>
+
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+ApplySquashing::ApplySquashing(Block header_)
+    : header(header_)
+{
+}
+
+Chunk ApplySquashing::add(Chunk && input_chunk)
+{
+    if (!input_chunk.hasChunkInfo())
+        return Chunk();
+
+    const auto *info = getInfoFromChunk(input_chunk);
+    append(info->chunks);
+
+    return std::move(accumulated_chunk);
+}
+
+void ApplySquashing::append(std::vector<Chunk> & input_chunks)
+{
+    accumulated_chunk = {};
+    std::vector<IColumn::MutablePtr> mutable_columns = {};
+    size_t rows = 0;
+    for (const Chunk & chunk : input_chunks)
+        rows += chunk.getNumRows();
+
+    {
+        auto & first_chunk = input_chunks[0];
+        Columns columns = first_chunk.detachColumns();
+        for (size_t i = 0; i < columns.size(); ++i)
+        {
+            mutable_columns.push_back(IColumn::mutate(std::move(columns[i])));
+            mutable_columns[i]->reserve(rows);
+        }
+    }
+
+    for (size_t i = 1; i < input_chunks.size(); ++i) // We've already processed the first chunk above
+    {
+        Columns columns = input_chunks[i].detachColumns();
+        for (size_t j = 0, size = mutable_columns.size(); j < size; ++j)
```

Comment:
for (auto & source_column: mutable_columns)
{
   mutable_columns[j]->insertRangeFrom(std::move(source_column), 0, source_column->size());
}